sdk/ai/ai-inference-rest/src/trace.ts

```diff
@@ -0,0 +1,360 @@
+            if (toolCall.function.arguments) {
+              toolCall.function.arguments = "";
+            }
+            toolCall.function.name = "";
+          });
+          content.tool_calls = toolCalls;
+        }
+      }
+
+      const toolMsg: ChatRequestToolMessage = message;
+      if (toolMsg.tool_call_id) {
+        content.id = toolMsg.tool_call_id;
+      }
+
+      span.addEvent(`gen_ai.${message.role}.message`, {
+        "gen_ai.system": INFERENCE_GEN_AI_SYSTEM_NAME,
+        "gen_ai.event.content": JSON.stringify(content),
+      });
+    }
+  });
+}
+
+/*
+* Add event to span.  Sample:
+{
+  name: 'gen_ai.choice',
+  attributes: {
+    'gen_ai.system': 'INFERENCE_GEN_AI_SYSTEM_NAME',
+    'gen_ai.event.content': '{"finish_reason":"tool_calls","index":0,"message":{"content":""}}'
+  },
+  time: [ 1725666881, 780608000 ],
+  droppedAttributesCount: 0
+}  
+*/
+function addResponseChatMessageEvent(span: TracingSpan, response: PathUncheckedResponse) {
+  response.body?.choices?.forEach((choice: ChatChoiceOutput) => {
+    let attributes;
+    let message: { content?: string; toolCalls?: Array<ChatCompletionsToolCallOutput> } = {};
+
+    if (choice.message.content) {
+      message.content = choice.message.content;
+    }
+    if (choice.message.tool_calls) {
+      message.toolCalls = choice.message.tool_calls;
+    }
+
+    if (!isContentRecordingEnabled()) {
+      message = JSON.parse(JSON.stringify(message));
+      message.content = "";
+      if (message.toolCalls) {
+        message.toolCalls.forEach((toolCall) => {
+          if (toolCall.function.arguments) {
+            toolCall.function.arguments = "";
+          }
+          toolCall.function.name = "";
+        });
+      }
+    }
+
+    const response = {
+      finish_reason: choice.finish_reason,
+      index: choice.index,
+      message,
+    };
+    attributes = {
+      "gen_ai.system": INFERENCE_GEN_AI_SYSTEM_NAME,
+      "gen_ai.event.content": JSON.stringify(response),
+    };
+
+    span.addEvent("gen_ai.choice", attributes);
+  });
+}
+
+export const tracingPolicy: PipelinePolicy = {
+  name: "tracingPolicyAI",
+  async sendRequest(pipelineRequest, next) {
+    const url = new URL(pipelineRequest.url);
+
+    // /// TODO: the code for streaming needs to be clean up.   We will implement tracing for streaming later
+    // if (!traceClient || request.body?.stream == true) {
+    //   return methodToTrace();
+    // }
+    const operationName = getOperationName(url.pathname);
+    const name = `${operationName}`.trim();
+    const { span, updatedOptions } = traceClient.startSpan(name, {
+      tracingOptions: pipelineRequest.tracingOptions,
+    });
+    pipelineRequest.tracingOptions = updatedOptions.tracingOptions;
+    onStartTracing?.(span, [{ ...pipelineRequest, headers: {} }, name, pipelineRequest.url]);
+
+    return traceClient
+      .withContext(updatedOptions.tracingOptions.tracingContext, () => next(pipelineRequest))
```

Comment:
_strongly_ recommend using `async/await` here instead of chaining - I left it as close as possible to the original code to avoid confusion. But the upside is in your package you do you! 😄